### PR TITLE
[1712] Support interface to add support users

### PR DESCRIPTION
**Pull request by asmega**
### Context

- https://trello.com/c/lkhFVwWu/1712-third-line-support-can-add-other-support-users
- Save the need for a developer to onboard additional support users

### Changes proposed in this pull request

- Third line support users can:
  - view privileged users: support users, computacenter users and mno users
  - can also add other support users - these are restricted to certain domains for the email addresses
  - revoke all privileges of one of these users

### Screenshots

New section on support homepage 
image

Index view
image

Adding another user
image

### Guidance to review

- Sign in as non third line support
- Cannot see or access this areas
- Sign in as third line support
- Should be able to see and access new area
- Can see existing privileged users
- Can add privileged users
- Can remove privileges of users

### CODE REVIEW

File: app/form_objects/support/privileged_user_form.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  include ActiveModel::Model
+
+  attr_accessor :full_name, :email_address, :privileges
+
+  validates :full_name, presence: { message: 'Enter the name of the user' }
+  validates :email_address, presence: { message: 'Enter the email address of the user' },
+                            format: { with: /\A.*@computacenter.com|.*@digital.education.gov.uk|.*@education.gov.uk\z/,
+                                      message: 'Email address must be end with @computacenter.com, @digital.education.gov.uk or @education.gov.uk' }
+  validate :validate_email_not_taken
+  validate :validate_atleast_one_privilege
```
Comment (nigel-lowry) [inline app/form_objects/support/privileged_user_form.rb:11]: I'd probably add an underscore so the method becomes `validate_at_least_one_privilege`

File: app/form_objects/support/privileged_user_form.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    end
+  end
+
+  def validate_atleast_one_privilege
+    if privileges.reject(&:blank?).empty?
+      errors.add :privileges, :at_least_one, message: 'Select at least one privilege to apply'
+    end
+  end
+
+  def validate_email_not_taken
```
Comment (nigel-lowry) [inline app/form_objects/support/privileged_user_form.rb:40]: Can't this use the standard uniqueness validation?

File: app/form_objects/support/privileged_user_form.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  validate :validate_atleast_one_privilege
+  validate :validate_acceptable_privileges
+
+  def create_user!
+    User.create!(full_name: full_name, email_address: email_address, is_support: is_support?, is_computacenter: is_computacenter?)
+  end
+
+private
+
+  def is_support?
```
Comment (nigel-lowry) [inline app/form_objects/support/privileged_user_form.rb:20]: If these predicate methods were renamed to something like `support?` then we could use RSpec syntactic sugar on the assertions (see other comment in pull request).

File: spec/form_objects/support/privileged_user_form_spec.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    subject(:form) { described_class.new(full_name: 'some body', email_address: 'user@digital.education.gov.uk', privileges: %w[support computacenter]) }
+
+    it 'sets up privileges' do
+      expect {
+        form.create_user!
+      }.to change(User, :count).by(1)
+
+      user = User.last
+
+      expect(user.is_support?).to be_truthy
```
Comment (nigel-lowry) [inline spec/form_objects/support/privileged_user_form_spec.rb:82]: If we changed the method name, the assertion could become `expect(user).to be_support`

**Review by nigel-lowry — APPROVED** [review]

File: spec/form_objects/support/privileged_user_form_spec.rb (hunk 4 above)
Comment (asmega) [reply spec/form_objects/support/privileged_user_form_spec.rb:82]: unfortunately these are already present as db column names. if we want to change these i think it needs to be part of a different ticket and addressed holistically as i touches lots of different parts. but yes i agree we can have these switched over

File: app/form_objects/support/privileged_user_form.rb (hunk 3 above)
Comment (asmega) [reply app/form_objects/support/privileged_user_form.rb:20]: See my other comment. this is private so damage is limited but can be changed when we tackle the whole problem

File: app/form_objects/support/privileged_user_form.rb (hunk 2 above)
Comment (asmega) [reply app/form_objects/support/privileged_user_form.rb:40]: This is not a an ActiveRecord model, so I can't use the standard uniqueness validator as it doesn't hook up to a data source.

**Review by nigel-lowry — APPROVED** [review]

**Pull request merged**